Close in stages - waiting for releases

martinsumner opened the pull request:
[Related issue](https://github.com/martinsumner/leveled/issues/410)

Have a consistent approach to closing the inker and the penciller - so that the close can be interrupted by releasing of snapshots.  Then any unreleased snapshots are closed before shutdown - with a 10s pause to give queries a short opportunity to finish.

This should address some issues, primarily seen (but very rarely) in test whereby post-rebuild destruction of parallel AAE keystores cause the crashing of aae_folds.

The primary benefit is to stop an attempt to release a snapshot that has in fact already finished, from causing a crash of the database on normal stop - which in Riak may prevent other closing actions from completing normally.  This was primarily an issue when shutdown is delayed by an ongoing journal compaction job.

martinsumner committed:
Close in stages - waiting for releases

Have a consistent approach to closing the inker and the penciller - so that the close can be interrupted by releasing of snapshots.  Then any unreleased snapshots are closed before shutdown - with a 10s pause to give queries a short opportunity to finish.

This should address some issues, primarily seen (but very rarely) in test whereby post-rebuild destruction of parallel AAE keystores cause the crashing of aae_folds.

The primary benefit is to stop an attempt to release a snapshot that has in fact already finished does not cause a crash of the database on normal stop.  this was primarily an issue when shutdown is delayed by an ongoing journal compaction job.

martinsumner committed:
Boost default test budget for EQC

ThomasArts commented on src/leveled_inker.erl:
```diff
@@ -778,8 +773,41 @@ handle_cast({remove_logs, ForcedLogs}, State) ->
+handle_cast({complete_shutdown, ShutdownType, From}, State) ->
+    lists:foreach(
+        fun(SnapPid) -> ok = ink_close(SnapPid) end,
+        lists:filter(
```
Matter of taste, but I find this harder to read than a list comprehension... up to you
In this case, keep the foreach, but filter in the comprehension:
```
[ Snap || {Snap, _TS, _SQN} <- State#state.registered_snapshots, is_process_alive(Snap) ]
```
On a more Meta level... should ink_close(SnapPid) not be able to deal with Pids that are not alive?
After all, there is always the possibility that the snapshot pids terminates just between the check and asking it to close?

What happens now if the pid is when you come to close is still that this may occasionally crash.
But since you know you're shutting down, you could just catch the `ink_close` instead of matching on `ok`. After all, why stop the cleanup if one of the snapshots exists or does not return ok?

martinsumner replied:
I've changes to catch the noproc scenario, which removes the need for the filter.

ThomasArts commented on src/leveled_inker.erl:
```diff
@@ -866,13 +895,6 @@ start_from_file(InkOpts) ->
-%% @doc
-%% Shutdown any snapshots before closing the store
-shutdown_snapshots(Snapshots) ->
-    lists:foreach(fun({Snap, _TS, _SQN}) -> ok = ink_close(Snap) end,
```
This original code failed when a snapshot already was closed or did respond with something else than `ok`.
However, in a cleanup, what does it give you when you crash the gen_server loop because you cannot cleanup?

ThomasArts commented on src/leveled_penciller.erl:
```diff
@@ -1156,7 +1157,36 @@ handle_cast({remove_logs, ForcedLogs}, State) ->
+    {noreply, State};
+handle_cast({complete_shutdown, ShutdownType, From}, State) ->
+    lists:foreach(
+        fun(Snap) -> ok = pcl_close(Snap) end,
```
symmetric argument to ink_close in inker

martinsumner replied:
Changed as with Inker

ThomasArts commented on src/leveled_pmanifest.erl:
```diff
@@ -88,22 +89,23 @@
+                    blooms :: dict:dict()
                     }).      
 
+-type fake_pid() :: pid_a1|pid_a2|pid_a3|pid_a4.
```
is this to please dialyzer because some tests use a symbolic representation of pids?
If so, I were to fix those tests by defining variables:
`PidA1 = spawn(fun() -> ok end)`
and use such fake pids instead. At least they have the right type and they can be compared on equality, but nothing more.

martinsumner replied:
Fixed as suggested

ThomasArts commented on src/leveled_pmanifest.erl:
```diff
@@ -538,10 +540,9 @@ mergefile_selector(Manifest, LevelIdx, {grooming, ScoringFun}) ->
-                            min_snapshot_sqn =
-                                PencillerManifest#manifest.min_snapshot_sqn}.
+    ClerkManifest#manifest{
+        snapshots = PencillerManifest#manifest.snapshots,
```
Eh... should that not be a list and therefore never match the atom `snapshots`?

If so, this seems untested code

martinsumner replied:
It isn't matching.  There are two versions of the manifest, both manifest records.  We want to take everything from the Clerk's manifest going forward (which has just completed a file compaction event), but still use the snapshot information taken from the Pencillers manifest (as it is the Penciller not the Clerk that has been receiving updates to this).  So it takes the Clerks Manifest and just sets the snapshot information on the record to be that of the Penciller.

ThomasArts reviewed (commented):
Some "up to choice" comments, but have a serious look at pmanifest line 544

martinsumner committed:
Update test to use correct type

martinsumner committed:
Update following review

Avoid filtering out exited PIDs when closing snapshots by catching the exit exception when the Pid is down
